### ISSUE TITLE
Supporting ocaml versions without classic OMP

### ISSUE THREAD

**Issue by emillon** [ocaml-ppx/ocamlformat#1460]
**TL;DR:** *we won't be able to support ocaml 4.12 with future versions of OMP. I think we should use ppxlib and I'm offering to port ocamlformat.*

Hi,

Starting with version 2.0.0, ocaml-migrate-parsetree (OMP) has been considerably simplified. More info and pointers can be found in [this message](https://discuss.ocaml.org/t/ocaml-migrate-parsetree-2-0-0/5991).

In a nutshell, "classic" OMP (1.x.y) has several components:

1. versioned AST modules (e.g. `Ast_408.Parsetree` is a copy of `Parsetree` as of ocaml 4.08)
2. migration functions between versions
3. versioned copies of compiler-libs APIs (e.g. a version of `Ast_mapper that operates on 4.08 ASTs)
4. a driver to register and run transformations

And OMP 2 only contains 1 and 2. In ocamlformat, we use 1, 2 and 3.

We need to take action because there won't be a classic OMP release that supports OCaml 4.12, which is due by the end of year.

An option would be to make a library that replicates classic OMP for us, but this has various drawbacks. First, it's too much maintenance work, and it would require backporting knowledge about future versions of ocaml back into older ocamlformat releases.

As suggested by the announcement linked above, I think that our best option is to use `ppxlib`. It supports only the latest ocaml version, but it gets updated early in the ocaml release cycle. It provides APIs similar to those of `compiler-libs` for this AST version. These are often higher level, so it's easier to write a function that collects all doc-comments on a structure for example. It is necessary to follow `ppxlib` updates to benefit from AST updates, but [`ppxlib` maintainers have offered to send PRs to do this for us](https://discuss.ocaml.org/t/ppx-omp-2-0-0-and-next-steps/6231).

As an aside, `ocamlformat` is a bit of a different use case because we are not emitting binary ASTs, unlike ppx rewriters do. In theory, it shouldn't matter which version of ocaml we're running on. In practice, it does because we are always using the parser from compiler-libs, which brings some problems in practice (on older ocaml versions, upwards migrations run, so there can be some discrepancies compared to a migrationless path). This is something we can solve in the future by packaging a parser corresponding to the AST version in ppxlib. This parser can be used instead of the one from compiler-libs, isolating ocamlformat from the version it has been compiled under. There are some rough edges with that (such as when we retokenize source code), and it's mostly independent from the issue of what to do with OMP.

To be clear, I'm offering to do the work, but I'd like to make sure we're on the same page before doing that.
Let me know what you think @gpetiot @jberdine @ceastlund.

**Comment by ceastlund** [comment]
This sounds fine to me.

**Comment by gpetiot** [comment]
It sounds like a good plan indeed.

**Comment by jberdine** [comment]
At a high level, I think that this is a good plan. The functionality of omp to enable intentional use of an old version of the parsetree, and map newer ones back to it, is something that we don't need: supporting the constructs of the new compiler version is always reasonably high priority. So trading the versioned Ast modules for a single pretty fresh one is good.

Regarding point 3, it would be good to record what modules we use from omp that will need to be looked at. There is currently some hacking to get `Warning` to work in < 4.08 due to the necessary module not being versioned by omp, and I wonder if we will end up needing more of that sort of workaround. This isn't an objection to the plan, I'm just saying that I'd like a better idea of the map of APIs we use and where they will come from.

Regarding vendoring the parser, I agree that it is a separate question, but one thing that I think we should keep in mind is that the current situation where the check that the parsed ast after formatting is equivalent to the one before formatting gives a notably stronger guarantee when the parser used is the compiler's parser. Losing that property might be necessary in practice though. We could still enable each ocamlformat executable to know which compiler version's parser it uses, and to be able to report that or complain if a mismatch is detected.

(We could merge https://github.com/ocaml-ppx/ocamlformat/issues/305 into this one.)

**Comment by emillon** [comment]
Nice, thanks all for your input.

Last time I had a look, we use at least `Ast_mapper` in delimited places (`Cmts` and `Normalize`), and have few uses of `Ast_helper` (easy to replace).

The `Warning` compat layer we have is indeed not provided by OMP (so we'll have to keep it), but I'd say it's something we have to use because we use the parser from compiler-libs. So it would have to be part of the ppxlib parser API. The versioning of `Parser.token` is in the same situation.

**Comment by hhugo** [comment]
How much effort would it be to not depend on compiler-libs, omp, ppxlib all together ? I suspect that with small synchronisation scripts to vendor/import part of the compiler codebase, the cost would not be too high.

**Comment by emillon** [comment]
It's difficult to evaluate. I think OMP can be similarly described as a git import and some scripts, but that was a ton of work for the maintainers in the long run.

**Comment by hhugo** [comment]
Most of the omp maintenance work is related to provide migration functions between ast versions et keep compatibility with old compiler-libs. All this would be unnecessary in the ocamlformat case.
